centrallix/expression/exp_functions.c

```diff
@@ -3978,368 +4151,232 @@ int exp_fn_nth(pExpression tree, pParamObjects objlist, pExpression i0, pExpress
+	return -1;
 	}
-
-    if ((i0->DataType != DATA_T_STRING) || (i1->DataType != DATA_T_STRING) || (i2 && i2->DataType != DATA_T_INTEGER))
+    
+    /** Extract string param. **/
+    pExpression maybe_str = check_ptr(tree->Children.Items[0]);
+    if (maybe_str->Flags & EXPR_F_NULL)
 	{
-		mssError(1,"EXP","lev_compare() requires two string and one optional integer parameters");
-		return -1;
+	tree->Flags |= EXPR_F_NULL;
+	tree->DataType = DATA_T_STRING;
+	return 0;
 	}
-	
-	exp_fn_levenshtein(tree, objlist, i0, i1, i2);
-	//!!! I am not checking for errors here, because IN THEORY we have two strings... if we don't, big uh-oh.
-	int lev_dist = tree->Integer;
-	
-	int length1 = strlen(i0->String);
-	int length2 = strlen(i1->String);
-
-	double clamped_dist = 1.0;
-
-	if (length1 == 0 || length2 == 0) //empty string
+    const char* str = check_ptr(maybe_str->String);
+    const size_t str_len = strlen(str);
+    if (str_len == 0u)
 	{
-		clamped_dist = 0.5;	
-	} 
-	else //normal case 
-	{
-		int max_len = (length1 > length2) ? length1 : length2;
-		clamped_dist = ((double) lev_dist) / max_len;
-	
-		if (abs(length1-length2) == lev_dist)  //only inserts. Maybe substring.
-		{
-			clamped_dist /= 2;
-		}
-		
-		//use max_field_width if it was provided as a sensible value. If not, don't use it.
-		double max_field_width = i2?(i2->Integer):0;
-		if (max_field_width && max_field_width >= max_len) {
-			double mod = (lev_dist + max_field_width * 3/4) / max_field_width; 
-			if (mod < 1) { //don't make clamped_dist bigger
-				clamped_dist *= mod;
-			}
-		}
-	}
-	
-	
-	tree->DataType = DATA_T_DOUBLE;
-	tree->Types.Double = 1.0 - clamped_dist;
+	tree->String = "";
+	tree->DataType = DATA_T_STRING;
 	return 0;
-}
-
-// This is the size of the vector table. It is also used in calculating the table indices.
-const int EXP_VECTOR_TABLE_SIZE = 251;
-
-/*
- * hash_char_pair
- * This method creates an vector table index based a given character pair. The characters are represented 
- * as their ASCII code points.
- *
- * Parameters:
- * 	num1 : first ASCII code point (double)
- * 	num2 : second ASCII code point (double)
- *
- * Returns:
- * 	vector table index (integer)
- */
-int exp_fn_i_hash_char_pair(double num1, double num2)
-    {
-    int func_result = round(((num1 * num1 * num1) + (num2 * num2 * num2)) * ((num1+1)/(num2+1))) -1;
-    return func_result % EXP_VECTOR_TABLE_SIZE;	
+	}
+    
+    /** Compute DoubleMetaphone. **/
+    char* primary = NULL;
+    char* secondary = NULL;
+    meta_double_metaphone(str, &primary, &secondary);
+    
+    /** Process result. **/
+    const size_t result_length = strlen(primary) + 1u + strlen(secondary) + 1u;
+    char* result = check_ptr(nmSysMalloc(result_length * sizeof(char*)));
+    if (result == NULL) return -1;
+    sprintf(result, "%s%c%s", primary, CA_BOUNDARY_CHAR, secondary);
+    
+    /** Return the result. **/
+    tree->String = result;
```

Comment:
Review the existing usage patterns of tree->String.  The Expression structure contains a preallocated buffer for small strings, to avoid extra memory allocations in many (most?) cases when dealing with strings.

If result_length < 64, it can be put in the StringBuf and Alloc needs to be set to 0.  If >= 64, allocate it and set Alloc to 1.  If Alloc is 1 before setting the new value, nmSysFree the old String value.